File: LeetCode_Python/n-queens_II.py
class Solution(object):
    def backtrack(self, total, row, cols, diag1, diag2, n):
        if row == n:
            total += 1
            return total

        for col in range(n):
            if col in cols or (row - col) in diag1 or (row + col) in diag2:
                continue

            cols.add(col)
            diag1.add(row - col)
            diag2.add(row + col)

            total = self.backtrack(total, row + 1, cols, diag1, diag2, n)

            cols.remove(col)
            diag1.remove(row - col)
            diag2.remove(row + col)

        return total

    def totalNQueens(self, n):
        total = self.backtrack(0, 0, set(), set(), set(), n)
        return total

n = 5
sol = Solution()
print(sol.totalNQueens(n))

File: LeetCode_C/n-queens_II.c
#include <stdio.h>

void backtrack(int n, int row, int cols[], int diagonals[], int antiDiagonals[], int *totalSolutions) {
    if (row == n) {
        (*totalSolutions)++;
        return;
    }

    for (int col = 0; col < n; col++) {
        if (cols[col] || diagonals[row - col + n - 1] || antiDiagonals[row + col]) {
            continue;
        }

        cols[col] = 1;
        diagonals[row - col + n - 1] = 1;
        antiDiagonals[row + col] = 1;

        backtrack(n, row + 1, cols, diagonals, antiDiagonals, totalSolutions);
       
        cols[col] = 0;
        diagonals[row - col + n - 1] = 0;
        antiDiagonals[row + col] = 0;
    }
}

int totalNQueens(int n) {
    int cols[n];
    int diagonals[2 * n - 1];
    int antiDiagonals[2 * n - 1];
    for (int i = 0; i < n; i++) cols[i] = 0;
    for (int i = 0; i < 2 * n - 1; i++) {
        diagonals[i] = 0;
        antiDiagonals[i] = 0;
    }

    int totalSolutions = 0;
    backtrack(n, 0, cols, diagonals, antiDiagonals, &totalSolutions);
    return totalSolutions;
}

int main() {
    int n = 5;
    int result = totalNQueens(n);
    printf("%d", result);
    return 0;
}